File: src/sc_neurocore/accel/go/services/sc_sigma_delta_accumulator.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li

package services

import (
	"errors"
	"math"
)

// SCSigmaDeltaAccumulatorState contains the frozen bipolar project state.
type SCSigmaDeltaAccumulatorState struct{ Sigma, VThreshold float64 }

// NewSCSigmaDeltaAccumulator constructs the project default.
func NewSCSigmaDeltaAccumulator() *SCSigmaDeltaAccumulatorState {
	return &SCSigmaDeltaAccumulatorState{0, 1}
}

// Valid reports whether project state is finite and configured.
func (s SCSigmaDeltaAccumulatorState) Valid() bool {
	return !math.IsNaN(s.Sigma) && !math.IsInf(s.Sigma, 0) && !math.IsNaN(s.VThreshold) && !math.IsInf(s.VThreshold, 0) && s.VThreshold > 0
}

// Step advances one atomic signed accumulator transition.
func (s *SCSigmaDeltaAccumulatorState) Step(current float64) (int, error) {
	if !s.Valid() || math.IsNaN(current) || math.IsInf(current, 0) {
		return 0, ErrSCSigmaDeltaInvalid
	}
	sigma := s.Sigma + current
	if math.IsNaN(sigma) || math.IsInf(sigma, 0) {
		return 0, ErrSCSigmaDeltaInvalid
	}
	event := 0
	if sigma >= s.VThreshold {
		sigma -= s.VThreshold
		event = 1
	} else if sigma <= -s.VThreshold {
		sigma += s.VThreshold
		event = -1
	}
	s.Sigma = sigma
	return event, nil
}

// ErrSCSigmaDeltaInvalid reports an invalid project transition.
var ErrSCSigmaDeltaInvalid = errors.New("SC sigma-delta accumulator state or input is invalid")

File: src/sc_neurocore/accel/rust/safety/sc_sigma_delta_accumulator.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li

//! Fail-closed retained project bipolar accumulator.

/// Frozen SC state and threshold.
#[derive(Debug, Clone)]
pub struct SCSigmaDeltaAccumulatorNeuron {
    pub sigma: f64,
    pub v_threshold: f64,
}
impl SCSigmaDeltaAccumulatorNeuron {
    /// Construct the historical default profile.
    pub fn new() -> Self {
        Self {
            sigma: 0.0,
            v_threshold: 1.0,
        }
    }
    /// Advance one atomic project transition.
    pub fn step(&mut self, current: f64) -> Result<i32, &'static str> {
        if !current.is_finite() || !validate_sc_sigma_delta_accumulator(self) {
            return Err("invalid SC SigmaDelta accumulator");
        }
        let mut sigma = self.sigma + current;
        if !sigma.is_finite() {
            return Err("SC SigmaDelta candidate is non-finite");
        }
        let event = if sigma >= self.v_threshold {
            sigma -= self.v_threshold;
            1
        } else if sigma <= -self.v_threshold {
            sigma += self.v_threshold;
            -1
        } else {
            0
        };
        self.sigma = sigma;
        Ok(event)
    }
    /// Clear the accumulator.
    pub fn reset(&mut self) {
        self.sigma = 0.0;
    }
}
impl Default for SCSigmaDeltaAccumulatorNeuron {
    fn default() -> Self {
        Self::new()
    }
}
/// Validate retained project state.
pub fn validate_sc_sigma_delta_accumulator(s: &SCSigmaDeltaAccumulatorNeuron) -> bool {
    s.sigma.is_finite() && s.v_threshold.is_finite() && s.v_threshold > 0.0
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn signed_transition_matches_project() {
        let mut s = SCSigmaDeltaAccumulatorNeuron::new();
        assert_eq!(s.step(3.25), Ok(1));
        assert_eq!(s.sigma, 2.25);
        assert_eq!(s.step(-4.5), Ok(-1));
    }
}
